include/onnxruntime/core/framework/op_kernel_context.h

```diff
@@ -175,6 +175,7 @@ class OpKernelContext {
   const OrtValue* GetInputMLValue(int index) const;
   const OrtValue* GetImplicitInputMLValue(int index) const;
   OrtValue* GetOutputMLValue(int index);
+  Status SetOutputMLValue(int index, const OrtValue& ort_value);
 
```

Comment:
If this is only required for training can we ifdef it based on that being enabled? Not sure what would happen if someone set this during inferencing w.r.t. the allocation plan and potential re-use of buffers.

Reply:
Possibly also needs some comments about correct usage as well. I'm assuming it's dependent on certain things happening in the allocation planner which are dependent on the kernel def saying it uses external outputs. But not sure anything can enforce that was the case when SetOutputMLValue is being called.

---
In reply to: [595596868](https://github.com/microsoft/onnxruntime/pull/6973#discussion_r595596868) [](ancestors = 595596868)

Reply:
Moved all SetOutputMLValue under #ifdef ENABLE_TRAINING